src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,294 @@
+        std::accumulate(logProbs.begin() + wordIndex.begin, logProbs.begin() + wordIndex.end, float(0.0)) /
+        wordIndex.size());
+  }
+
+  const float sentenceScore =
+      std::accumulate(std::begin(wordScores), std::end(wordScores), float(0.0)) / wordScores.size();
+
+  return {wordScores, subwordToWords(wordIndexes, target, sentenceIdx), sentenceScore};
+}
+
+LogisticRegressorQualityEstimator::Matrix::Matrix(const size_t rowsParam, const size_t colsParam)
+    : rows(rowsParam), cols(colsParam), data_(rowsParam * colsParam) {}
+
+LogisticRegressorQualityEstimator::Matrix::Matrix(Matrix&& other)
+    : rows(other.rows), cols(other.cols), data_(std::move(other.data_)) {}
+
+const float& LogisticRegressorQualityEstimator::Matrix::at(const size_t row, const size_t col) const {
+  return data_[row * cols + col];
+}
+
+float& LogisticRegressorQualityEstimator::Matrix::at(const size_t row, const size_t col) {
+  return data_[row * cols + col];
+}
+
+LogisticRegressorQualityEstimator::LogisticRegressorQualityEstimator(Scale&& scale, std::vector<float>&& coefficients,
+                                                                     const float intercept)
+    : scale_(std::move(scale)),
+      coefficients_(std::move(coefficients)),
+      intercept_(intercept),
+      coefficientsByStds_(coefficients_.size()) {
+  ABORT_IF(scale_.means.size() != scale_.stds.size(), "Number of means is not equal to number of stds");
+  ABORT_IF(scale_.means.size() != coefficients_.size(), "Number of means is not equal to number of coefficients");
+
+  // Pre-compute the scale operations for the linear model
+  for (int i = 0; i < coefficients_.size(); ++i) {
+    coefficientsByStds_[i] = coefficients_[i] / scale_.stds[i];
+    constantFactor_ += coefficientsByStds_[i] * scale_.means[i];
+  }
+}
+
+LogisticRegressorQualityEstimator::LogisticRegressorQualityEstimator(LogisticRegressorQualityEstimator&& other)
+    : scale_(std::move(other.scale_)),
+      coefficients_(std::move(other.coefficients_)),
+      intercept_(std::move(other.intercept_)),
+      coefficientsByStds_(std::move(other.coefficientsByStds_)),
+      constantFactor_(std::move(other.constantFactor_)) {}
+
+LogisticRegressorQualityEstimator LogisticRegressorQualityEstimator::fromAlignedMemory(
+    const AlignedMemory& alignedMemory) {
+  LOG(info, "[data] Loading Quality Estimator model from buffer");
+
+  const char* ptr = alignedMemory.begin();
+  const size_t blobSize = alignedMemory.size();
+
+  ABORT_IF(blobSize < sizeof(Header), "Quality estimation file too small");
+  const Header& header = *reinterpret_cast<const Header*>(ptr);
+
+  ABORT_IF(header.magic != BINARY_QE_MODEL_MAGIC, "Incorrect magic bytes for quality estimation file");
+  ABORT_IF(header.lrParametersDims <= 0, "The number of lr parameter dimension cannot be equal or less than zero");
+
+  const size_t numLrParamsWithDimension = 3;  // stds, means and coefficients
+  const size_t numIntercept = 1;
```

Comment:
Looks like both of these should become a data member (constexpr or const) of the class?

Reply:
These can be a compile-time constant (`constexpr`). Also I'm guessing there's a `numFeatures=4` somewhere which can be a `constexpr`.

Reply:
Commit 9c68d67